### PR TITLE
Write pickle to file-like without intermediate in-memory buffer

### LINKED ISSUE #34244: ENH: Support out-of-band pickling (protocol 5)
#### Is your feature request related to a problem?

It would be nice if Pandas objects supported [pickle's protocol 5 for out-of-band serialization]( https://www.python.org/dev/peps/pep-0574/ ). This would allow the underlying data to be captured in `PickleBuffer`s (specialized `memoryview`). For libraries using pickle's protocol 5 to transmit data over the wire, this would allow for zero-copy data transmission.

#### Describe the solution you'd like

Pandas objects implement `__reduce_ex__` and if the `protocol` argument is `5` or greater, they construct `PickleBuffer`s out of any data arguments.

#### API breaking implications

NA as it should be possible to fallback to existing behavior for older pickle protocols. Users have to actively opt-in at a higher level API (through pickle) to see any effect.

#### Describe alternatives you've considered

NA

#### Additional context

This would be useful in libraries that support distributed dataframes ;)

### PR DESCRIPTION
**Pull request by ig248**
Before this change, calling `pickle.dumps()` created an in-memory byte buffer, negating the advantage
of zero-copy pickle protocol 5. After this change, `pickle.dump` writes directly to open file(-like),
cutting peak memory in half in most cases.

Profiling was done with pandas@master and python 3.8.5

Related issues:
- https://github.com/pandas-dev/pandas/issues/34244

image

## Update: ASV results
`$ asv continuous -f 1.1 origin/master HEAD -b pickle` yields:
```
       before           after         ratio
     [58f74686]       [3c1ee270]
                      <zero-copy-pickle>
-           91.9M            80.1M     0.87  io.pickle.Pickle.peakmem_write_pickle

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

- [ ] closes #34244
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
nice, can you add a memory asv and show the results of this & a whatsnew note for 1.2 perf section.

**Comment by jreback** [comment]
also to actually close this issue i think would be good to add some tests that explicity set the pickle protocal to 5

**Comment by ig248** [comment]
> also to actually close this issue i think would be good to add some tests that explicity set the pickle protocal to 5

Isn't that covered by py38 tests where pickle.HEIGHEST_PROTOCOL (==5) is used by default?

**Comment by jreback** [comment]
> > also to actually close this issue i think would be good to add some tests that explicity set the pickle protocal to 5
> 
> Isn't that covered by py38 tests where pickle.HIGHEST_PROTOCOL (==5) is used by default?

probably but let's make it explict as well (e.g. parameterize on 4,5,highest when PY38 or greater)

**Comment by ig248** [comment]
Added ASV peakmem_ benchmark. The benchmark currently uses an ~11MB dataframe - I have no experience with ASV to tell how reliable memory benchmarking is for small-ish objects, any advice appreciated.

**Comment by jreback** [comment]
> Added ASV peakmem_ benchmark. The benchmark currently uses an ~11MB dataframe - I have no experience with ASV to tell how reliable memory benchmarking is for small-ish objects, any advice appreciated.

it *should* show a diff here in any event. what kind of results are you seeing?

File: pandas/tests/io/test_pickle.py [hunk 1]
```diff
@@ -155,28 +157,44 @@ def test_pickles(current_pickle_data, legacy_pickle):
...
+
+named_pickle_writers = {
+    "python": python_pickler,
+    "pandas_proto_default": pd.to_pickle,
+    "pandas_proto_highest": functools.partial(
+        pd.to_pickle, protocol=pickle.HIGHEST_PROTOCOL
+    ),
+    "pandas_proto_4": functools.partial(pd.to_pickle, protocol=4),
+}
+if sys.version_info >= (3, 8):
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py:179]: import from pandas.compat PY38

File: pandas/tests/io/test_pickle.py [hunk 2]
```diff
@@ -155,28 +157,44 @@ def test_pickles(current_pickle_data, legacy_pickle):
...
+    named_pickle_writers.update(
+        {
+            "pandas_proto_5": functools.partial(pd.to_pickle, protocol=5),
+        }
+    )
+pickle_writer_ids, pickle_writers = zip(*named_pickle_writers.items())
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py:185]: alternativey you can use something like: 

```pytest.param(...., marks=pytest.mark.skipif(not PY38, reason=...)```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/io/test_pickle.py (hunk 1 above)
Comment (ig248) [reply pandas/tests/io/test_pickle.py:179]: I assume we should test protocol=5 for all 3.8+, not just `if PY38`?

File: pandas/tests/io/test_pickle.py (hunk 2 above)
Comment (ig248) [reply pandas/tests/io/test_pickle.py:185]: `pytest.parm` is more readable - however, again do we want to skip for 3.9+?

**Comment by ig248** [comment]
> > Added ASV peakmem_ benchmark. The benchmark currently uses an ~11MB dataframe - I have no experience with ASV to tell how reliable memory benchmarking is for small-ish objects, any advice appreciated.
> 
> it _should_ show a diff here in any event. what kind of results are you seeing?

Yep, added results above - `peakmem_write_pickle` shows the 11MB drop nicely - in practice the relative gain is obv much more significant than shown for larger dataframes.

File: pandas/tests/io/test_pickle.py (hunk 1 above)
Comment (ig248) [reply pandas/tests/io/test_pickle.py:179]: oh sorry, PY38 is defined as >=(3, 8) already

File: doc/source/whatsnew/v1.2.0.rst [hunk 3]
```diff
@@ -313,6 +313,7 @@ Performance improvements
   avoiding creating these again, if created on either. This can speed up operations that depend on creating copies of existing indexes (:issue:`36840`)
 - Performance improvement in :meth:`RollingGroupby.count` (:issue:`35625`)
 - Small performance decrease to :meth:`Rolling.min` and :meth:`Rolling.max` for fixed windows (:issue:`36567`)
+- Reduced peak memory usage in DataFrame.to_pickle() when using protocol=5 in python 3.8+
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:316]: can you make this ``` :meth:`DataFrame.to_pickle` ``` and put protocol=5 in double back-ticks so they will render; also add the issue number

File: pandas/tests/io/test_pickle.py [hunk 4]
```diff
@@ -155,28 +156,43 @@ def test_pickles(current_pickle_data, legacy_pickle):
...
+def python_unpickler(path):
+    with open(path, "rb") as fh:
+        fh.seek(0)
+        return pickle.load(fh)
+
+
+@pytest.mark.parametrize(
+    "pickle_writer",
+    [
+        pytest.param(python_pickler, id="python"),
```
Comment (jreback) [inline pandas/tests/io/test_pickle.py:173]: you can omit the `pytest.param` for all of these, except the xfailed one (if you are ok with the id that is generated) i believe

Comment (jreback) [reply pandas/tests/io/test_pickle.py:173]: but this is pretty clear too, so up to you if it works

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm ping on green. can you post the results of the asv (in the top of the PR, just edit it)

File: pandas/tests/io/test_pickle.py (hunk 4 above)
Comment (ig248) [reply pandas/tests/io/test_pickle.py:173]: IMHO it looks less clear without:
```
pandas/tests/io/test_pickle.py::test_round_trip_current[python_pickler] PASSED                                                                                                               [ 20%]
pandas/tests/io/test_pickle.py::test_round_trip_current[to_pickle] PASSED                                                                                                                    [ 40%]
pandas/tests/io/test_pickle.py::test_round_trip_current[pickle_writer2] PASSED                                                                                                               [ 60%]
pandas/tests/io/test_pickle.py::test_round_trip_current[pickle_writer3] PASSED                                                                                                               [ 80%]
pandas/tests/io/test_pickle.py::test_round_trip_current[pandas_proto_5] PASSED  
```
- so I would prefer to leave it with custom `id`s

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @ig248 just resolved the conflict and can merge on green.

**Comment by jreback** [comment]
thanks @ig248 very nice!

**Pull request merged**